src/math/GraphUtil.cpp

```diff
@@ -0,0 +1,107 @@
+}
+
+std::vector<stdcxx::optional<unsigned long >>& GraphUtil::ConnectedComponentsComputationResult::getComponentNumber() {
+    return m_componentNumber;
+}
+
+const std::vector<unsigned long>& GraphUtil::ConnectedComponentsComputationResult::getComponentSize() const {
+    return m_componentSize;
+}
+
+std::vector<unsigned long>& GraphUtil::ConnectedComponentsComputationResult::getComponentSize() {
+    return m_componentSize;
+}
+
+GraphUtil::ConnectedComponent::ConnectedComponent(unsigned long size) :
+    m_size(size) {
+}
+
+unsigned long GraphUtil::ConnectedComponent::getOrderedNumber() const {
+    return m_orderedNumber;
+}
+
+unsigned long GraphUtil::ConnectedComponent::getSize() const {
+    return m_size;
+}
+
+GraphUtil::ConnectedComponent& GraphUtil::ConnectedComponent::setOrderedNumber(unsigned long orderedNumber) {
+    m_orderedNumber = orderedNumber;
+    return *this;
+}
+
+GraphUtil::ConnectedComponentsComputationResult GraphUtil::computeConnectedComponents(const std::vector<std::vector<unsigned long>>& adjacencyList) {
+    std::vector<stdcxx::optional<unsigned long>> componentNumber(adjacencyList.size());
+    unsigned long c = 0UL;
+    std::vector<unsigned long> componentSize(adjacencyList.size(), 0UL);
+    for (unsigned long v = 0UL; v < adjacencyList.size(); v++) {
+        if (!componentNumber[v].is_initialized()) {
+            computeConnectedComponents(v, c, componentSize, adjacencyList, componentNumber);
+            c++;
+        }
+    }
+
+    // sort components by size
+    std::vector<std::shared_ptr<std::pair<unsigned long, stdcxx::optional<unsigned long>>>> components;
+    components.reserve(c);
+    std::vector<std::shared_ptr<std::pair<unsigned long, stdcxx::optional<unsigned long>>>> orderedComponents;
+    orderedComponents.reserve(c);
+
+    for (unsigned long i = 0UL; i < c; i++) {
+        auto comp = std::make_shared<std::pair<unsigned long, stdcxx::optional<unsigned long>>>(componentSize[i], stdcxx::optional<unsigned long>());
+        components.push_back(comp);
+        orderedComponents.push_back(comp);
+    }
+    auto compare = [](const std::shared_ptr<std::pair<unsigned long, stdcxx::optional<unsigned long>>>& o1, const std::shared_ptr<std::pair<unsigned long, stdcxx::optional<unsigned long>>>& o2) {
+        return o2->first < o1->first;
+    };
+    std::sort(orderedComponents.begin(), orderedComponents.end(), compare);
+    for (unsigned long i = 0UL; i < orderedComponents.size(); i++) {
+        orderedComponents[i]->second = i;
+    }
+
+    componentSize = std::vector<unsigned long>(orderedComponents.size());
+    for (const auto& cc : orderedComponents) {
+        componentSize[*cc->second] = cc->first;
+    }
+    for (unsigned long i = 0UL; i < componentNumber.size();) {
+        const auto& cc = components[*componentNumber[i]];
+        componentNumber[i] = cc->second;
+        ++i;
+    }
+    return ConnectedComponentsComputationResult(componentNumber, componentSize);
+}
```

Comment:
Need to review this algorithm again: don't sure this is totally equivalent to the java implementation and not sure we have to use shared_ptr here

Reply:
It looks like shared pointer are necessary to make the code more similar to Java, because, both vectors point on the same data in java, like in C++:
```
        // sort components by size
        ConnectedComponent[] components = new ConnectedComponent[c];
        ConnectedComponent[] orderedComponents = new ConnectedComponent[c];
        for (int i = 0; i < c; i++) {
            ConnectedComponent comp = new ConnectedComponent(componentSize[i]);
            components[i] = comp;
            orderedComponents[i] = comp;
        }
```